README.md
# shinybg

[![R-CMD-check](https://github.com/TileDB-Inc/shinybg/workflows/R-CMD-check/badge.svg)](https://github.com/TileDB-Inc/shinybg/actions)

## Overview

Run and manage Shiny applications as background processes.

With *shinybg* you can:

* launch a Shiny application without locking your current R session
* run multiple applications at once
* embed application in Jupyter notebooks

## Installation

*shinybg* is currently under active development and is not yet available on CRAN. You can install the development version from GitHub:

```r
remotes::install_github("tiledb-inc/shinybg", remotes::github_release())
```

## Usage

### Basic app management

Create two instances of the same app on ports 3001 and 3002.

```r
app <- system.file("apps/sever-info-app.R", package = "shinybg")
app1 <- runBackgroundApp(appFile = app, port = 3001)
app2 <- runBackgroundApp(appFile = app, port = 3002)
```

### Jupyter integration

You can test out embedding Shiny applications in a Jupter notebook using the provided [Dockerfile][]. Pre-built images are available on [Docker Hub](https://hub.docker.com/repository/docker/tiledb/shinybg) or you can build it yourself by running:

```sh
docker build -t tiledb/shinybg:latest .
```

Next, run the container, forwarding ports 8888 for the Jupyter server and 3000 for the embedded Shiny app.

```sh
docker run --rm \
  -p 8888:8888 \
  -p 3000:3000 \
  -e JUPYTER_ENABLE_LAB=yes \
  -v $PWD/dev:/home/jovyan/work \
  tiledb/shinybg:latest
```

Within the Jupyter environment, launch an R-kernel Jupyter notebook and paste the following chunk into the first cell:

```r
library(shiny)
library(shinybg)

renderShinyApp(
    appFile = system.file("apps/sever-info-app.R", package = "shinybg")
)
```

## Get in touch

- [@tiledb](https://twitter.com/tiledb)
- [TileDB Community Slack](https://tiledb-community.slack.com)
- [TileDB Support Forum](https://forum.tiledb.com)



NEWS.md
# shinybg v0.1.6

* `runBackgroundApp()` and `renderShinyApp()` gain arguments for specifying the location of the app's stdout and stderr streams (#12)
* Adds exported functions for app management
* Update Docker hub repo name
# shinybg v0.1.5

* Rename package to *shinybg*

# tiledbJupyterShiny 0.1.4

* Adds `AppManager` class for managing background shiny apps
* New `runBackgroundApp()` function for launching apps as background tasks without rendering an IFrame
* New demo app for testing
* Included Jupyter notebook provides an overview of the app manager's functionality

# tiledbJupyterShiny 0.1.3

* `renderShinyApp() now detects when JupyterLab is running inside kubernetes and uses jupyter-server-proxy's endpoint to serve the shiny app  (#4)
* Improved look of IFrames by removing the border (#5) and matching the height of the viewport (#6)

# tiledbJupyterShiny 0.1.2

* Added a `NEWS.md` file to track changes to the package.
* Apps are now launched as background tasks using the *callr* package.
* We now verify an app is up and running before rendering the IFrame


R/manage.R
#' List background Shiny applications
#' @return a list of [`callr::r_process`] objects named for the port the Shiny
#'    app is listening on.
#' @family app management functions
#' @export
list_apps <- function() {
    app_manager$apps
}

#' Retrieve a background Shiny application
#' @param port TCP port the app is listening on (e.g., 3000)
#' @family app management functions
#' @export
get_app <- function(port) {
  app_manager$retrieve_app(port)
}

#' Kill a background Shiny application
#' @inheritParams get_app
#' @family app management functions
#' @export
kill_app <- function(port) {
  app_manager$kill_app(port)
}

#' Kill all background Shiny applications
#' @family app management functions
#' @export
kill_all_apps <- function() {
  app_manager$kill_all_apps()
}

#' View background Shiny application
#'
#' Open an application in an external browser or within the configured RStudio
#' viewer
#'
#' @inheritParams get_app
#' @return (Invisibly) The application's URL.
#' @family app management functions
#' @export
view_app <- function(port) {
  app_manager$view_app(port)
}


R/renderShinyApp.R
#' Render Shiny Application in a Jupyter Notebook
#'
#' @inheritParams runBackgroundApp
#' @inheritParams shiny::shinyApp
#' @importFrom IRdisplay display_html
#' @importFrom pingr is_up
#' @export

renderShinyApp <- function(
  ui = NULL,
  server = NULL,
  appFile = NULL,
  appDir = NULL,
  port = 3000,
  stdout = "|",
  stderr = "|"
) {

  # hardcode host for rendering shiny within iframe
  host <- "0.0.0.0"
  rproc <- runBackgroundApp(
    ui = ui,
    server = server,
    appFile = appFile,
    appDir = appDir,
    port = port,
    host = host,
    stdout = stdout,
    stderr = stderr
  )

  while(!pingr::is_up(destination = host, port = port)) {
    if (!rproc$is_alive()) stop(rproc$read_all_error())
    Sys.sleep(0.05)
  }

  iframe_host <- getShinyHost(port)
  displayIframe(iframe_host)
}


#' Get Shiny Host for IFrame
#' If jupyter is running inside k8s-hub create a url to use jupyter-server-proxy
#' @noRd
getShinyHost <- function(port) {
  jupyterUser <- Sys.getenv("JUPYTERHUB_USER")
  jupyterApiURL <- Sys.getenv("JUPYTERHUB_API_URL")

  if (nzchar(jupyterUser) && nzchar(jupyterApiURL)) {
    host <- sprintf("/user/%s/proxy/%s/", jupyterUser, port)
    return(host)
  }
  host <- sprintf('http://127.0.0.1:%s', port)
  return(host)
}


displayIframe <- function(host) {
  html <- sprintf('<iframe src="%s" width="100%%" style="border: 0;height: calc(100vh - 70px);margin: 0;"></iframe>', host)
  display_html(html)
}


R/runBackgroundApp.R
#' Run Shiny App in a Background Process
#'
#' @param ui The UI definition of the app.
#' @param port The TCP port that the application should listen on (defaults to 3000).
#' @inheritParams shiny::runApp
#' @inheritParams shiny::shinyApp
#' @inheritParams callr::r_bg
#' @importFrom IRdisplay display_html
#' @return A [callr::r_process] object for the background Shiny app
#' @examples
#' \dontrun{
#' app <- system.file("apps/sever-info-app.R", package = "shinybg")
#' bg_app <- runBackgroundApp(appFile = app, port = 3005)
#' bg_app$kill()
#' }
#' @import shiny
#' @importFrom callr r_bg
#' @export

runBackgroundApp <- function(
  ui = NULL,
  server = NULL,
  appFile = NULL,
  appDir = NULL,
  port = getOption("shiny.port"),
  host = getOption("shiny.host", "127.0.0.1"),
  stdout = "|",
  stderr = "|"
) {
  if (!is.null(ui) || !is.null(server)) {
    app <- shiny::shinyApp(ui, server)
  } else if (!is.null(appFile)) {
    app <- shiny::shinyAppFile(appFile)
  } else if (!is.null(appDir)) {
    app <- shiny::shinyAppDir(appDir)
  } else {
    stop(
      "You must define either 'ui'/'server', 'appFile', or 'appDir'",
      call. = FALSE
    )
  }

  run_app <- function(appDir, host, port) {
    shiny::runApp(appDir, host = host, port = port)
  }
  args <- list(appDir = app, host = host, port = port)

  if (port %in% app_manager$list_ports()) {
    app_manager$kill_app(port)
  }

  app <- callr::r_bg(
    func = run_app,
    args = args,
    stdout = stdout,
    stderr = stderr
  )

  app_manager$register_app(port, app)
  app
}


R/renderShinyDashboard.R
# Hello, world!
#
# This is an example function named 'hello'
# which prints 'Hello, world!'.
#
# You can learn more about package authoring with RStudio at:
#
#   http://r-pkgs.had.co.nz/
#
# Some useful keyboard shortcuts for package authoring:
#
#   Install Package:           'Ctrl + Shift + B'
#   Check Package:             'Ctrl + Shift + E'
#   Test Package:              'Ctrl + Shift + T'

#' @importFrom IRdisplay display_html
#' @import shiny
#' @import future

port <- 3000

runServer <- function(ui) {
  options("shiny.port" = port)
  app <- shinyApp(
    ui = ui,
    options = list(host = "0.0.0.0"),
    server = function(input, output) {
      output$plot <- renderPlot({ hist(runif(input$n)) })
    }
  )

  runApp(app)
}

displayIframe <- function() {
  html <- sprintf('<iframe src="http://127.0.0.1:%s" width="100%%", height="800"></iframe>', port)
  display_html(html)
}

#' @export
renderShinyApp <- function(ui) {
  plan(multisession)
  future(runServer(ui))
  Sys.sleep(1)
  displayIframe()
}


R/appManager.R
#' @title AppManager
#' @description Manage shiny apps running in background tasks.
#' @param port TCP port the app is listening on (e.g., 3000)
#' @param verbose (logical) Should messages be printed to the console?
#' @export
#' @examples \dontrun{
#' # Make a stub
#' manager <- AppManager$new()
#' manager
#'
#' # Create background shiny apps
#' app <- system.file("apps/sever-info-app.R", package = "shinybg")
#' app1 <- runBackgroundApp(appFile = app, port = 3001)
#' app2 <- runBackgroundApp(appFile = app, port = 3002)
#'
#' # Register the apps
#' manager$register_app(port = 3001, app = app1)
#' manager$register_app(port = 3002, app = app2)
#'
#' # List all managed apps
#' manager
#' manager$list_ports()
#'
#' # Retrieve app by port
#' manager$retrieve_app(port = "3001")
#'
#' # Kill all apps
#' manager$kill_all_apps()
#' }
AppManager <- R6::R6Class(
  "AppManager",
  public = list(
    #' @field apps (list) list of running shiny apps named for their
    apps = list(),

    #' @description print method for the `AppManager` class
    #' @param x self
    #' @param ... ignored
    print = function(x, ...) {
      cat("<shinybg apps> ", sep = "\n")
      if (length(self$apps) == 0) {
        cat("..No managed apps..", sep = "\n")
      } else {
        cat(" Running apps", sep = "\n")
        ports <- names(self$apps)
        for (i in seq_along(self$apps)) {
          cat(ports[[i]], "|", self$apps[[i]]$format())
        }
      }
      invisible(self$apps)
    },

    #' @description Register an app
    #' @param app [callr::r_process] for a Shiny app
    #' @return nothing returned
    register_app = function(port, app) {
      stopifnot(inherits(app, "r_process"))
      port <- as.character(port)
      if (!is.null(self$apps[[port]])) {
        stop(
          paste("Another process is already registered on port", port),
          call. = FALSE
        )
      }
      self$apps[[port]] <- app
    },

    #' @description Retrieve a registered app
    #' @return [callr::r_process] for a Shiny app
    retrieve_app = function(port) {
      port <- private$check_port_exists(port)
      self$apps[[port]]
    },

    #' @description View background Shiny application
    #' @return (Invisibly) The application's URL.
    #' @importFrom utils browseURL
    view_app = function(port) {
      port <- private$check_port_exists(port)

      # TODO: should the app manager store the registered app's URL?
      url <- getShinyHost(port)

      # From shiny: http://0.0.0.0/ doesn't work on QtWebKit (i.e. RStudio viewer)
      if (grepl("0.0.0.0", url, fixed = TRUE)) {
        url <- sub("0.0.0.0", "127.0.0.1", url, fixed = TRUE)
      }

      launch.browser <- getOption('shiny.launch.browser', interactive())

      # Adapted from shiny::runapp() to respect configured RStudio viewer
      if (is.function(launch.browser)) {
        launch.browser(url)
      } else if (launch.browser) {
        utils::browseURL(url)
      } else {
        url <- NULL
      }
      invisible(url)
    },

    #' @description Kill an app's process and deregister it
    #' @return `TRUE` if the process was terminated, and `FALSE` if it was not
    kill_app = function(port, verbose = TRUE) {
      port <- private$check_port_exists(port)
      result <- self$apps[[port]]$kill()
      if (result) {
        if (verbose) {
          message(paste("Killed app registered on port", port))
        }
        self$apps[[port]] <- NULL
      }
      result
    },

    #' @description Kill all registered apps
    #' @return Nothing returned
    kill_all_apps = function(verbose = TRUE) {
      killed <- vapply(self$apps, function(x) x$kill(), FUN.VALUE = logical(1L))
      if (verbose) message(sprintf("Successfully killed %i apps", sum(killed)))
    },

    #' @description List ports used by registered apps
    #' @return Character vector with 0 or more elements
    list_ports = function() {
      names(self$apps)
    }
  ),

  private = list(
    check_port_exists = function(port) {
      port <- as.character(port)
      if (!port %in% names(self$apps)) {
        stop(
          paste("No app is registered for port", port),
          call. = FALSE
        )
      }
      port
    }
  )
)
